### DISCUSSION TITLE
Using readline instead of editline in Python and others

### DISCUSSION THREAD

**Discussion by dtwilliamson** [Homebrew/discussions#4632] [Everyday usage]
### Output of `brew config`

```shell
n/a
```

### Output of `brew doctor`

```shell
n/a
```

## Use readline instead of editline (MacOS)

This discussion provides a solution to the issue in **brew-core** _Control-r (search command history) doesn't work in Python@3.11_ [#118869](https://github.com/Homebrew/homebrew-core/issues/118869). Using these instructions you can disable editline and use [readline](https://linux.die.net/man/3/readline) in its place for command line editing in the Python REPL and other programs.

These instructions should work in other Unix-like/Linux OSes in addition to MacOS.

### Disable editline

Edit `~/.editrc` and decide what to remove. Then add:

    edit off

or prefix it with the application name or regex to limit which ones have editline disabled:

    python3:edit off

The only line I previously had in that file was `bind -e` which I removed.

You can provide multiple prefixed lines (or an appropriate regex) to have the command apply to additional applications.

### Install [`rlwrap`](https://github.com/hanslub42/rlwrap)

    brew install rlwrap

`rlwrap` has a bunch of options, see the output of:

    rlwrap --help

or the [man page](https://linux.die.net/man/1/rlwrap).

### Edit `~/.inputrc`

If you need to make any changes or additions to readline bindings you can edit this file. If you want to have configurations for particular programs wrap the settings in `$if` conditionals.

For system-global configuration, use `/etc/inputrc`. But note that a user-local configuration completely overrides the global one instead of being preferentially merged. However, you can include the global file (or others) using `$include`. Placement of those lines within a file controls precedence.

### Create aliases

In your `~/.zshrc` or `~/.bashrc` or other startup files (depending on how your shell is started), create aliases to make the use of `rlwrap` more convenient. For example:

    alias python3='rlwrap python3'

I keep my aliases in a separate file and `source` them in my main startup file. I use `~/bin/aliases` but `~/.aliases` is another suggestion.

You can also run programs on an ad-hoc basis without an alias by prefixing the program name with `rlwrap` followed by a space (but without the surrounding quotes seen in the `alias` command).

### Adjust your history files

For example, my Python history file was `~/.python_history` but the one created after using `rlwrap` is `~/.python3_history`. Instead of changing some configuration, I just renamed my old history so I could continue to use it:

    mv ~/.python_history  ~/.python3_history

### Enjoy!

<kbd>**Ctrl**</kbd><kbd>**r**</kbd>**your-previous-command**

### And that's not all...

Try it with MySQL and others, too!

---
cc @kinghuang

**Comment by kinghuang** [comment]
Thanks, I'll give this a try! I've encountered many other problems beyond just `control-r` with editline.
